### PR TITLE
fixes  #127

### LINKED ISSUE #127: KeyShare error: "server selected unsupported group" with HelloFirefox_*
Using uTLS to access azure servers I get a hanshake error *[server selected unsupported group](https://github.com/refraction-networking/utls/blob/862fe372d939c53bb10cb4fd11263f3766589892/handshake_client_tls13.go#L228)* with any of the Firefox client hello.

A simple code to reproduce it:
```go
package main

import (
        "log"
        "net"

        utls "github.com/refraction-networking/utls"
)

var (
        defaultClientHello = utls.HelloFirefox_Auto
        host               = "ajax.aspnetcdn.com"
)

func main() {
        rawConn, err := net.Dial("tcp", host+":443")
        if err != nil {
                log.Fatal("Dial error: ", err)
        }
        tlsConn := utls.UClient(rawConn, &utls.Config{
                ServerName: host,
        }, defaultClientHello)
        defer tlsConn.Close()

        if err = tlsConn.Handshake(); err != nil {
                log.Fatal("Handshake error: ", err)
        }
}
```

This works fine with HelloChrome_*, I assume is related to some of the extensions that the Firefox client hello claims to support. It works fine with [yawning's fork](https://gitlab.com/yawning/utls), I think because of [50663a1c918572d973339ce6fe174d1375cca76d](https://gitlab.com/yawning/utls/-/commit/50663a1c918572d973339ce6fe174d1375cca76d), but this fork has an incompatible license we can't just import this commit here :disappointed:.

### PR DESCRIPTION
**Pull request by fqrious**

### CODE REVIEW

File: handshake_client_tls13.go [hunk 1]
```diff
@@ -17,12 +17,27 @@ import (
...
-	ctx         context.Context
-	serverHello *serverHelloMsg
-	hello       *clientHelloMsg
-	ecdheParams ecdheParameters
+	c                    *Conn
+	ctx                  context.Context
+	serverHello          *serverHelloMsg
+	hello                *clientHelloMsg
+	ecdheParams          ecdheParameters
+	keySharesEcdheParams KeySharesEcdheParameters
```
Comment (gaukas) [inline handshake_client_tls13.go:40]:
```suggestion
	keySharesEcdheParams KeySharesEcdheParameters // [UTLS]
```

File: handshake_client_tls13.go [hunk 2]
```diff
@@ -17,12 +17,27 @@ import (
...
+	keymap[curveID] = params
+}
+func (keymap KeySharesEcdheParameters) GetEcdheParams(curveID CurveID) (params ecdheParameters, ok bool) {
+	params, ok = keymap[curveID]
+	return
+}
+func (keymap KeySharesEcdheParameters) GetPublicEcdheParams(curveID CurveID) (params EcdheParameters, ok bool) {
+	params, ok = keymap[curveID]
+	return
+}
```
Comment (gaukas) [inline handshake_client_tls13.go:32]: Please make sure to enclose changes to non `u_` files in `// [UTLS SECTION START]` `// [UTLS SECTION END]` for future reference.

File: handshake_client_tls13.go [hunk 3]
```diff
@@ -55,6 +70,11 @@ func (hs *clientHandshakeStateTLS13) handshake() error {
 		return errors.New("tls: server selected TLS 1.3 in a renegotiation")
 	}
 
+	// set echdheParams to what we received from server
+	if ecdheParams, ok := hs.keySharesEcdheParams.GetEcdheParams(hs.serverHello.serverShare.group); ok {
+		hs.ecdheParams = ecdheParams
+	}
```
Comment (gaukas) [inline handshake_client_tls13.go:76]: `// [UTLS SECTION START]` `// [UTLS SECTION END]`...

File: handshake_client.go [hunk 4]
```diff
@@ -219,14 +219,15 @@ func (c *Conn) clientHandshake(ctx context.Context) (err error) {
...
-			ecdheParams: ecdheParams,
-			session:     session,
-			earlySecret: earlySecret,
-			binderKey:   binderKey,
+			c:                    c,
+			ctx:                  ctx,
+			serverHello:          serverHello,
+			hello:                hello,
+			ecdheParams:          ecdheParams,
+			keySharesEcdheParams: make(KeySharesEcdheParameters, 2),
```
Comment (gaukas) [inline handshake_client.go:227]:
```suggestion
			keySharesEcdheParams: make(KeySharesEcdheParameters, 2), // [UTLS]
```

**Review by gaukas — COMMENTED** [review]
Looks legit. I suggested a few changes maybe you want to check out.

**Comment by fqrious** [comment]
I think I've resolved all suggestions

**Review by gaukas — APPROVED** [review]

**Pull request merged**